prov/verbs/src/windows/verbs_nd_ibv.c

```diff
@@ -0,0 +1,857 @@
+int ___ibv_query_port(struct ibv_context *context, uint8_t port_num,
+		      struct ibv_port_attr *port_attr)
+{
+	trace(__func__);
+	if ((!context) || (!port_attr)) {
+		errno = EINVAL;
+		return errno;
+	}
+
+	struct adapter_t *adapter = get_adapter_by_context(context);
+	if (!adapter) {
+		errno = ENODEV;
+		return errno;
+	}
+
+	memset(port_attr, 0, sizeof(struct ibv_port_attr));
+	port_attr->state = IBV_PORT_ACTIVE;
+	port_attr->max_mtu = IBV_MTU_4096;
+	port_attr->active_mtu = IBV_MTU_4096;
+	port_attr->gid_tbl_len = 0;
+	port_attr->max_msg_sz = adapter->info.MaxTransferLength;
+	port_attr->pkey_tbl_len = 0;
+	port_attr->lid = 0;
+	port_attr->sm_sl = 0;
+	port_attr->active_width = 2; // 4x
+	port_attr->active_speed = 32; // 25 Gbps
+	port_attr->phys_state =
+		5; // IB_PORT_PHYS_STATE_LINKUP from "infiniband/iba/ib_types.h"
+	port_attr->link_layer = IBV_LINK_LAYER_ETHERNET;
+
+	return FI_SUCCESS;
+}
+
+int ibv_query_gid(struct ibv_context *context, uint8_t port_num, int index,
+		  union ibv_gid *gid)
+{
+	trace(__func__);
+	gid->global.subnet_prefix =
+		0x80fe; // little endian representation of link-local address prefix.
+	gid->global.interface_id = 0;
+	return FI_SUCCESS;
+}
+
+int ibv_query_pkey(struct ibv_context *context, uint8_t port_num, int index,
+		   __be16 *pkey)
+{
+	trace(__func__);
+	*pkey = 0xffff;
+	return FI_SUCCESS;
+}
+
+struct ibv_pd *ibv_alloc_pd(struct ibv_context *context)
+{
+	trace(__func__);
+	if (!context) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	struct ibv_pd *pd = (struct ibv_pd *)calloc(1, sizeof(struct ibv_pd));
+	if (!pd) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	pd->context = context;
+
+	return pd;
+}
+
+int ibv_dealloc_pd(struct ibv_pd *pd)
+{
+	trace(__func__);
+	if (!pd) {
+		errno = EINVAL;
+		return errno;
+	}
+	free(pd);
+	return FI_SUCCESS;
+}
+
+struct ibv_comp_channel *ibv_create_comp_channel(struct ibv_context *context)
+{
+	trace(__func__);
+	if (!context) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	struct ibv_comp_channel *channel = (struct ibv_comp_channel *)calloc(
+		1, sizeof(struct ibv_comp_channel_vrb_nd));
+	if (!channel) {
+		errno = ENOMEM;
+		return NULL;
+	}
+	struct ibv_comp_channel_vrb_nd *channel_nd =
+		container_of(channel, struct ibv_comp_channel_vrb_nd, channel);
+
+	int ret = dlistfd_head_init(&channel_nd->q);
+	if (ret) {
+		free(channel);
+		errno = ret;
+		return NULL;
+	}
+	ofi_mutex_init(&channel_nd->q_lock);
+	channel_nd->num_pending = 0;
+	channel->context = context;
+	channel->fd = fd_signal_get(&channel_nd->q.signal);
+
+	return channel;
+}
+
+int ibv_destroy_comp_channel(struct ibv_comp_channel *channel)
+{
+	trace(__func__);
+	if (!channel) {
+		errno = EINVAL;
+		return errno;
+	}
+
+	struct ibv_comp_channel_vrb_nd *channel_nd =
+		container_of(channel, struct ibv_comp_channel_vrb_nd, channel);
+
+	while (channel_nd->num_pending) {
+		int ret = dlistfd_wait_avail(&channel_nd->q, -1);
+		if (ret == 1) {
+			ofi_mutex_lock(&channel_nd->q_lock);
```

Comment:
I'll admit the num_pending, q_lock, and q members may be a bit confusing.

When ibv_req_notify_cq is called, num_pending is incremented because there will eventually be an entry added to q, but not until the asynchronous notify completes either successfully or in error, e.g. cancelled during teardown.

When ibv_destroy_comp_channel is called, we need to wait until all asynchronous requests are completed and add an entry to q. Thus the check for num_pending > 0 and not q empty.

I think the lock for q is needed, because ibv_destroy_comp_channel will be competing with nd_cq_notify_event/nd_cq_notify_error for access to the queue. The former trying to drain it and the latter trying to add entries.

So, I think both num_pending and q are needed and the logic is correct. I am going to change num_pending into an atomic though as there could be contention for it.

This was a tricky spot though, so please check my math. ;-)

Reply:
If another thread is trying to acquire the mutex, then it could try to do so after mutex_destroy (followed by free) has been called below, resulting in an invalid memory access.  The protection here isn't clear to me.

What is num_pending counting?  What increments the value and what decrements the value do not seem to be related.

Reply:
Let me take another crack at explaining this to see if I can make it make sense. The context for all this is a completion channel.
```
ibv_create_comp_channel
    initialize queue
    initialize mutex
    initialize num_pending to 0

ibv_req_notify_cq
    starts an asynchronous call that will lead to a callback to either
        nd_cq_notify_event or
        nd_cq_notify_error
    increments num_pending indicating there is a new outstanding call

nd_cq_notify_event/nd_cq_notify_error
  lock mutex
  insert completion event into the queue
  unlock mutex
```
In the normal course of events `ibv_get_cq_event` will be either polled or called after an epoll wait on the channel's fd returns.
```
ibv_get_cq_event
  lock mutex
  remove completion event from the queue
  decrement num_pending
  unlock mutex
```
Now lets look at what occurs during a shutdown scenario. In this case there will be a situation where the queue is empty, but `num_pending` is 1 because there is an outstanding notification request that hasn't completed.

The client code will call `ibv_destroy_cq`, which will trigger any outstanding requests to complete with a cancelled status, and then `ibv_destroy_comp_channel`, which wants to free all the associated resources.

So this sets up a race between `nd_cq_notify_error` and `ibv_destroy_comp_channel` where we want to force a wait to allow the outstanding requests to complete and insert an entry into the queue.

Looping until `num_pending` is 0 does this.
```
ibv_destroy_comp_channel
    while num_pending > 0
        wait for non-empty queue
        lock mutex
        remove completion event form the queue
        decrement num_pending
        unlock mutex
    end loop
```

So, there can only be contention for the mutex and queue if `num_pending > 0`. Since we loop until `num_pendng == 0` we will never destroy the mutex and queue and free the memory if there might be contention.

So, the logic still seems sound to me, but there is always a chance I'm wrong. Clearly it's complicated given the back and forth. So, pick it apart and point out any shortcomings. Also, if you think there is a clearer way to achieve the same result please share.

Reply:
One issue is that num_pending isn't directly related to what it is intending to track, which is the number of outstanding overlapped notify operations.  Once nd_cq_notify_event/nd_cq_notify_error have been called, the overlapped operation is done.  That's where num_pending should be decremented.  Plus the name num_pending isn't descriptive enough for someone to figure out what is actually pending.

Another issue is that the behavior of ibv_req_notify_cq is to 'arm' the cq to generate an event on the next completion.  While the cq is armed, additional calls to ibv_req_notify_cq are essentially no-ops.  There should only be 1 event generated for a CQ until it is re-armed after an event has been generated, not 1 event per ibv_req_notify_cq call.

Similarly, we should only have at most 1 outstanding notify request.  So this could be simplified by adding the overlapped structure directly to the cq with a flag to indicate if it's in use.  Destroy could check this, and if it's in use, cancel the overlapped, wait for it to complete, then cleanup the event list.  It's been too long since I've looked at ND and overlapped operations that I don't remember how to cancel them, but I assume there's a way.

Reply:
The ND documentation for CompletionQueue::Notify says: 
> Multiple requests for notification can be outstanding for the same completion queue. All such requests will be completed by the next completion event on the specified completion queue. This is similar to a notification event that releases all waiting requests, rather than a synchronization event that releases, at most, one. This allows multiple threads to process completions in parallel.

I think this is why I implemented it as I did. That being said, I suppose I could boil it down to match the behavior you describe for the Linux ibverbs implementation since that is what we're trying to emulate.

However, even after that, I don't think it can be quite as simple as you describe. There can be a many to one relationship between CQs and completion channels. And the data structures point towards a CQ knowing its channel, but a channel not knowing its CQs.

The documentation for ibv_destroy_comp_channel mentions that it will fail if any CQs are still associated with it. I must not have taken notice of that, because I sure didn't attempt to implement it that way. I also think I recall seeing a refcnt member in ibv_comp_channel. So perhaps the best way is not to count outstanding notify requests, but use the refcnt member to return an error if it is non-zero and if it's zero, then proceed to drain the queue of any events and release all the resources.

Reply:
My comment above is relative to the cq, not the comp_channel.  My reference to 'destroy' is to ibv_destroy_cq.  For the comp_channel, yes, I believe you can use a simple refcnt for the number of associated cq's and fail if not zero.

What I was describing above is adding an overlapped structure to the cq, along with some flag.  When ibv_req_notify_cq () is called, mark the overlapped as in-use.  When nd_cq_notify_event/nd_cq_notify_error is called, mark the overlapped as available.  In ibv_destroy_cq, see if the overlapped is in-use, and if it is, cancel it and wait for the cancel to finish.